python/cudf/cudf/core/indexing.py

```diff
@@ -344,13 +374,16 @@ def _getitem_tuple_arg(self, arg):
                 return self._downcast_to_series(df, arg)
             return df
         else:
-            df = DataFrame()
-            for i, col in enumerate(columns_df._columns):
-                # need Series() in case a scalar is returned
-                df[i] = Series(col[arg[0]])
-
-            df.index = as_index(columns_df.index[arg[0]])
-            df.columns = columns_df.columns
+            if isinstance(arg[0], slice):
+                df = columns_df._slice(arg[0])
+            elif is_scalar(arg[0]):
+                index = arg[0]
+                if index < 0:
+                    index += len(columns_df)
+                df = columns_df._slice(slice(index, index + 1, 1))
+            else:
+                arg = (column.as_column(arg[0]), arg[1])
+                df = columns_df.take(arg[0])
```

Comment:
Go directly down to gather here, bypass take

Reply:
This is handling both boolean array and numerical indexing, else have to branch out again for both.

Reply:
Lets branch out and handle both. We have too much layering of different APIs where we should make them all consistently just dispatch down to `slice`, `mask`, `gather`, etc.